src/digits/ff31.rs

```diff
@@ -0,0 +1,1373 @@
+                type Output = Monty;
+                #[inline]
+                fn sub(mut self, rhs: Monty) -> Monty {
+                    self -= rhs;
+                    self
+                }
+            }
+
+            impl SubAssign for Monty {
+                #[inline]
+                fn sub_assign(&mut self, other: Monty) {
+                    let a = &mut self.limbs;
+                    let needs_add =
+                        $classname::sub_assign_limbs_if(a, other.limbs, ConstantBool(1));
+                    $classname::add_assign_limbs_if(a, PRIME, needs_add);
+                }
+            }
+
+            impl PartialOrd for Monty {
+                #[inline]
+                fn partial_cmp(&self, other: &Monty) -> Option<Ordering> {
+                    self.limbs.const_ordering(&other.limbs)
+                }
+            }
+
+            impl Zero for Monty {
+                #[inline]
+                fn zero() -> Self {
+                    Monty {
+                        limbs: [0u32; NUMLIMBS],
+                    }
+                }
+
+                #[inline]
+                fn is_zero(&self) -> bool {
+                    self.limbs.const_eq0().0 == 1
+                }
+            }
+
+            impl One for Monty {
+                #[inline]
+                fn one() -> Self {
+                    let mut ret = Monty {
+                        limbs: [0u32; NUMLIMBS],
+                    };
+                    ret.limbs[0] = 1u32;
+                    ret
+                }
+
+                #[inline]
+                fn is_one(&self) -> bool {
+                    self.limbs.const_eq(Self::one().limbs).0 == 1
+                }
+            }
+
+            impl $classname {
+                ///Square the value. Same as a value times itself, but slightly more performant.
+                #[inline]
+                pub fn square(&self) -> $classname {
+                    let doublesize = $classname::mul_limbs_classic(&self.limbs, &self.limbs);
+                    $classname {
+                        limbs: $classname::reduce_barrett(&doublesize),
+                    }
+                }
+                #[inline]
+                pub fn to_monty(self) -> Monty {
+                    Monty { limbs: self.limbs } * Monty {
+                        limbs: MONTRSQUARED,
+                    }
+                }
+
+                #[inline]
+                pub fn normalize_assign_little(&mut self) {
+                    let new_limbs = $classname::normalize_little_limbs(self.limbs);
+                    self.limbs = new_limbs;
+                }
+
+                ///Take the extra limb and incorporate that into the existing value by modding by the prime.
+                /// This normalize should only be used when the input is at most
+                /// 2*p-1. Anything that might be bigger should use the normalize_big
+                /// options, which use barrett.
```

Comment:
Would be nice to note somewhere the cost/benefit that was done of when to use Monty vs Barrett...